### PR TITLE
Vdbe threaded code

### LINKED ISSUE #4212: sql: implement threaded interpretation for VDBE
Overall, problem is discussed here: https://github.com/tarantool/tarantool/issues/3330

Implementation of threaded interpretation is quite simple. However, it requires support of computed goto's by compiler. Most of mainstream compilers (gcc, clang, icc) provide that feature.

There are several examples of implementation:
SQLite https://github.com/AlexKashuba/SQLite_JIT/blob/master/versions/sqlite_disp/jitsrc/vdbe.c#L646
PostgreSQL https://github.com/postgres/postgres/blob/master/src/backend/executor/execExprInterp.c#L117

Plan is following:
1. Patch mkopcodec.sh and mkopcodeh.sh to produce array of goto addresses - it must be organised in the same order as corresponding OP_ values are defined.
2. Define macros DISPATCH/CASE to hide internal implementation of dispatching: if compiler doesn't support precomputed labels, then we should use common switched technique.
3. Replace break and case keywords with mentioned macros.
4. Benchmark results using TPC-H set of queries.

### PR DESCRIPTION
**Pull request by AnaNek**
Tarantool uses SQL VM with conventional interpretation technique: 
loop over instruction, inside which dispatching switch-case executes logic of current opcode.

In switch-based interpreter each instruction requires two jumps:
1. one indirect jump to the branch handling the current instruction
2. one direct jump from there to the main loop.

It would be better to avoid the second one, by jumping directly to the code handling the next instruction. 
This is called threaded code.

To measure the percentage of time spent in the VDBE in switch-based technique
and threaded code technique a query with recursion was used:

```sql
WITH RECURSIVE temp (n, sum) AS
(SELECT 0, 0
UNION ALL
SELECT n+1, (n+1)+sum FROM temp
WHERE n < 1000000)
SELECT * FROM temp;
```

**The percentage of time spent in the VDBE**

* switch-based technique: ~83%

* threaded code technique: ~76%

Closes #4212

### CODE REVIEW

File: src/box/sql/vdbe.c [hunk 1]
```diff
@@ -466,7 +466,182 @@ int sqlVdbeExec(Vdbe *p)
 			printf("VDBE Trace:\n");
 	}
 #endif
-	for(pOp=&aOp[p->pc]; 1; pOp++) {
+
+#define DISPATCH() \
+do{ \
+pOp++; \
```
Comment (Korablev77) [inline src/box/sql/vdbe.c]: Fix indentation (do block should be tabbed); place spaces before/after {; add a comment

File: src/box/sql/vdbe.c [hunk 2]
```diff
@@ -467,12 +467,110 @@ int sqlVdbeExec(Vdbe *p)
...
 #define DISPATCH() \
 do{ \
+VDBE_TRACE(); \
 pOp++; \
-goto *dispatch_table[pOp->opcode]; \
+CHECK_OPCODE(); \
+NEXT(pOp->opcode); \
 }while(0)
 
+#define START_EVAL() \
```
Comment (Korablev77) [inline src/box/sql/vdbe.c:567]: This macro seems to be redundant: just inline the code

File: src/box/sql/vdbe.c [hunk 3]
```diff
@@ -467,12 +467,110 @@ int sqlVdbeExec(Vdbe *p)
 	}
 #endif
 
+#if defined(SQL_DEBUG)
+
+#define VDBE_TRACE() \
```
Comment (Korablev77) [inline src/box/sql/vdbe.c:472]: I do not clearly understand why you introduced this macro..Why not simply inline/put it into function?
Macros almost always are considered to be evil (hard to read&patch)..

**Review by igormunkin — CHANGES_REQUESTED** [review]
@AnaNek, thanks for the patch! I agree with @Korablev77 regarding `VDBE_TRACE` and `CHECK_OPCODE` macros (I also see you've dropped all related comments, AFAICS). Please also consider my comments below.

Regarding the commit message: I guess you can describe the patch in a more verbose way. It's also worth to leave the link to our Wiki where your GSoC report can be found.

**Comment by Totktonada** [comment]
> To measure the percentage of time spent in the VDBE in switch-based technique
and threaded code technique a query with recursion was used:
>
> ```sql
> WITH RECURSIVE temp (n, sum) AS
> (SELECT 0, 0
> UNION ALL
> SELECT n+1, (n+1)+sum FROM temp
> WHERE n < 1000000)
> SELECT * FROM temp;
> ```

How about storing of the benchmarking code somewhere in the `perf/` directory?

**Comment by decalek** [comment]
@AnaNek, Hi Anastasia,

I was very happy to realize that someone started digging vdbe dispatch inefficiencies in practice!

As further optimization step (possibly with bigger magnitude), you might be interested in potential refactoring of vdbe as [Jittery VM][3], completely resolving #3330.

[Jitter][1] is a relatively [new][2] toolkit for building of efficient VMs, which supports traditional switch, threaded and something which the author coined as "**no-threading**" modes. AFAIU, the latter works as runtime generation of whole machine code programs, like a JIT, but instead own code generation for all ISAs, jitter at runtime reorders and mangles chunks of code generated by the C compiler in buildtime (the project supports GCC, clang, few more compilers (partially) and several architectures).

The usual "hello world" test (fib 44) on my machine, complete with these times:

* 0m6.851s: [fibo.c](http://git.ageinghacker.net/jitter/tree/example-vms/jitterlisp/examples/fibo.c)
* 0m6.894s: fennel-1.0 with luajit-2.1.b3
* 0m9.457s: chez, branch main (Chez is known as sophisticated compiler and one of the fastest scheme implementations)
* 0m12.712s: [jitterlisp](http://git.ageinghacker.net/jitter/tree/example-vms/jitterlisp/examples/fibo.lisp)

With this example VM you can see the generated assembly with
```
(disassemble fibo)
```
in the code or repl, the other two samples have distinct --disassemble switch on the command line.

The only jitter user so far is [GNU Poke][3] - the author Luca Saiu, besides the direct support and consulting of the GNU Poke project, don't bother too much with the marketing :-). It seems that current jitter discussion is enclosed in small group of people inside the GNU community.

Besides performance, vdbe as Jittery VM would bring #3050 for free (the toolkit generates VM assembly parser/pretty printer), which will be a convenience for eventual future development of more QL frontends over vdbe (e.g. datalog, SQL/GQL[5], etc.).

And there is one possible positive outcome for the very broad userbase of the SQLite project too: Once tarantool gained significant performance boost leveraging this approach, a [Lua GCC compiler plugin][4] could be contributed, which rewrites the source of the stock SQLite VDBE as Jittery VM. Many organizations (Linux distributors, Mozilla, Facebook and others heavy SQLite users) will be more than happy to use such (non-fork) elegant improvement.

Kind Regards,
Alek

[1]: http://ageinghacker.net/projects/jitter/
[2]: https://archive.org/details/jitter-presentation--2021-03-25
[3]: https://git.savannah.gnu.org/cgit/poke.git/tree/libpoke/pvm.jitter
[4]: https://peter.colberg.org/gcc-lua
[5]: https://www.gqlstandards.org/

**Comment by AnaNek** [comment]
I'm having problems with `checkpatch`.  The errors are related to macros. Some errors relate to the unwanted use of `goto` in a macro:

```
ERROR:MACRO_WITH_FLOW_CONTROL: Macros with flow control statements should be avoided
[87](https://github.com/tarantool/tarantool/actions/runs/3674262862/jobs/6212314363#step:4:98)
#388: FILE: src/box/sql/vdbe.c:495:
[88](https://github.com/tarantool/tarantool/actions/runs/3674262862/jobs/6212314363#step:4:99)
+#define NEXT(opcode)    goto *dispatch_table[opcode]
```

But the point of the task is to use `goto` in SQL vdbe. @locker , could you please take a look and tell me if I can ignore `checkpatch` errors?

File: extra/mkdispatchtableh.sh [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+while read line; do
+    case "$line" in
+    '#define OP_'*)
+        IFS=" "
+        set -- $line
+        IFS="$newline"
+        name="${2#_}"
+        case "$name" in
+        'OP_NotUsed_'*)
+            printf '%s,\n' "NULL" ;;
```
[the remark this answers is not in the document]
Comment (AnaNek) [reply extra/mkdispatchtableh.sh]: Fixed.

File: extra/mkdispatchtableh.sh [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+# This shell script scans the opcodes.h file (which is itself generated by
+# another shell script) and uses the information gleaned to create the
+# dispatch_table.h file.
+#
+# dispatch_table.h contains array of label addresses for opcodes.
+
+set -eu  # Strict shell (w/o -x and -o pipefail)
+set -f   # disable pathname expansion
+
+printf '%s' "static void* dispatch_table[] = {"
```
[the remark this answers is not in the document]
Comment (AnaNek) [reply extra/mkdispatchtableh.sh]: Fixed.

File: .gitignore [hunk 6]
```diff
@@ -161,6 +161,7 @@ src/box/sql/parse.h
 src/box/sql/parse.c
 src/box/sql/opcodes.h
 src/box/sql/opcodes.c
+src/box/sql/dispatch_table.h
```
[the remark this answers is not in the document]
Comment (AnaNek) [reply .gitignore]: Done.

File: src/box/CMakeLists.txt [hunk 7]
```diff
@@ -225,6 +225,10 @@ if(CMAKE_BUILD_TYPE STREQUAL "Debug")
 endif()
 add_definitions(-DSQL_TEST=1)
 
+if(WITHOUT_COMPUTED_GOTO STREQUAL "ON")
+  add_definitions(-DSQL_WITHOUT_COMPUTED_GOTO=1)
```
[the remark this answers is not in the document]
Comment (AnaNek) [reply src/box/CMakeLists.txt]: Done.

File: src/box/sql/vdbe.c [hunk 8]
```diff
@@ -4473,9 +4513,11 @@ case OP_SetSession: {
  * This opcode records information from the optimizer.  It is the
  * the same as a no-op.  This opcodesnever appears in a real VM program.
  */
-default: {          /* This is really OP_Noop and OP_Explain */
+
+EXECUTE(OP_Noop)
+EXECUTE(OP_Explain) {          /* This is really OP_Noop and OP_Explain */
```
[the remark this answers is not in the document]
Comment (AnaNek) [reply src/box/sql/vdbe.c]: Done.

File: src/box/sql/vdbe.c [hunk 9]
```diff
@@ -821,8 +869,6 @@ case OP_String8: {         /* same as TK_STRING, out2 */
 		goto too_big;
 	}
 	assert(rc == 0);
-	/* Fall through to the next case, OP_String */
```
[the remark this answers is not in the document]
Comment (AnaNek) [reply src/box/sql/vdbe.c:784]: Fixed.

File: src/box/sql/vdbe.c [hunk 10]
```diff
@@ -1236,10 +1282,8 @@ case OP_BuiltinFunction0: {
 	pOp->p4type = P4_FUNCCTX;
 	pOp->p4.pCtx = pCtx;
 	pOp->opcode = OP_BuiltinFunction;
-	/* Fall through into OP_BuiltinFunction */
```
[the remark this answers is not in the document]
Comment (AnaNek) [reply src/box/sql/vdbe.c]: Outdated.

File: src/box/sql/vdbe.c [hunk 11]
```diff
@@ -4473,9 +4513,11 @@ case OP_SetSession: {
  * This opcode records information from the optimizer.  It is the
  * the same as a no-op.  This opcodesnever appears in a real VM program.
  */
-default: {          /* This is really OP_Noop and OP_Explain */
+
+EXECUTE(OP_Noop)
+EXECUTE(OP_Explain) {          /* This is really OP_Noop and OP_Explain */
 	assert(pOp->opcode==OP_Noop || pOp->opcode==OP_Explain);
```
[the remark this answers is not in the document]
Comment (AnaNek) [reply src/box/sql/vdbe.c]: Deleted.

File: src/box/sql/vdbe.c [hunk 12]
```diff
@@ -4231,10 +4273,8 @@ case OP_AggStep0: {
 	pOp->p4type = P4_FUNCCTX;
 	pOp->p4.pCtx = pCtx;
 	pOp->opcode = OP_AggStep;
-	/* Fall through into OP_AggStep */
```
[the remark this answers is not in the document]
Comment (AnaNek) [reply src/box/sql/vdbe.c]: Outdated.

File: src/box/sql/vdbe.c [hunk 13]
```diff
@@ -3218,15 +3262,13 @@ case OP_Last: {        /* jump */
...
-case OP_SorterSort:    /* jump */
-case OP_Sort: {        /* jump */
+EXECUTE(OP_SorterSort)    /* jump */
+EXECUTE(OP_Sort) {        /* jump */
 #ifdef SQL_TEST
 			sql_sort_count++;
 			sql_search_count--;
 #endif
 			p->aCounter[SQL_STMTSTATUS_SORT]++;
-			/* Fall through into OP_Rewind */
```
[the remark this answers is not in the document]
Comment (AnaNek) [reply src/box/sql/vdbe.c:3292]: Fixed.

File: src/box/CMakeLists.txt [hunk 14]
```diff
@@ -225,6 +225,10 @@ if(CMAKE_BUILD_TYPE STREQUAL "Debug")
 endif()
 add_definitions(-DSQL_TEST=1)
 
+if(WITHOUT_COMPUTED_GOTO STREQUAL "ON")
```
[the remark this answers is not in the document]
Comment (AnaNek) [reply src/box/CMakeLists.txt]: Fixed.

File: src/box/sql/vdbe.c (hunk 3 above)
Comment (AnaNek) [reply src/box/sql/vdbe.c]: Fixed.

File: src/box/sql/vdbe.c (hunk 2 above)
Comment (AnaNek) [reply src/box/sql/vdbe.c]: Fixed.

File: src/box/sql/vdbe.c (hunk 1 above)
Comment (AnaNek) [reply src/box/sql/vdbe.c]: Done.

**Comment by locker** [comment]
> But the point of the task is to use `goto` in SQL vdbe. @locker , could you please take a look and tell me if I can ignore `checkpatch` errors?

checkpatch isn't run against the master branch so yes you can ignore its warning.

I'd rather not remove this warning from checkpatch, because using control-flow statements in macros **is** usually confusing. TBH the code looks weird to me with switch-case replaced with macros, but as long as @ImeevMA is okay with it, I don't mind.

**Comment by ImeevMA** [comment]
We are currently figuring out the future of VDBE, so I'll be back here later. Appointed for myself.

**Comment by ImeevMA** [comment]
Hi, @AnaNeck! Thanks for the PR, but I don’t plan to allow to pass it yet, because at the moment I don’t see a normal way to support the new code. PR suggests using quite a lot of macros and supporting two types of VDBE - using streaming implementation and using switch-case, which greatly increases the complexity of supporting VDBE.

**Comment by igormunkin** [comment]
Discussed with @ImeevMA offline and decided to re-open this until #8100 is resolved.

**Comment by ImeevMA** [comment]
We have decided to close this PR for now. It will be reopened if needed.

**Pull request closed without merging**